### PR TITLE
Fix large operators in cambria math

### LINKED ISSUE #3114: Integral symbol is too small in displayed equations when using Cambria Math font
### Description

I tried to use Cambria Math font with the following code:
```typst
Inline $integral_a^b f(x)=F(b)-F(a)$ and displayed
$ integral_a^b f(x)=F(b)-F(a) $

#show math.equation: set text(
  font: ("Cambria Math"),
)

Inline $integral_a^b f(x)=F(b)-F(a)$ and displayed
$ integral_a^b f(x)=F(b)-F(a) $
```
but the integral symbol in displayed equations looked too small with Cambria Math:

typst

The following is the same equation in Microsoft Word:

word

### Reproduction URL

_No response_

### Operating system

Windows

### Typst version

- [X] I am using the latest version of Typst

### PR DESCRIPTION
**Pull request by Enter-tainer**
fix #3114

TLDR: cambria math has problems and we have to workaround for it.

### CODE REVIEW

**Comment by Dherse** [comment]
Do we really want to fix more upstream issues directly in Typst? 🤔 
Could we instead package a fixed version of the font?

I know we're already doing a lot of overly convoluted workarounds for PDF and (to some degree) SVG, but this feels... a step too far?

Just a thought, overall I am not necessarily against this change, more playing devil's advocate.

Also, if we have workaround that are font-specific, I think we should have a better defined API like callbacks in a macro:
```rs
fallback! {
    fonts,
    "Cambria Math" => cambria_math_workaround(...),
    _ => ...
}
```

That way we can build on it for other problematic fonts that surely exist.

**Comment by Enter-tainer** [comment]
> Do we really want to fix more upstream issues directly in Typst? 🤔

It's complicated. Basically it is microsoft add some new parameters to otf math which doesn't exist in texbook appendix G. And only microsoft's math font actually using that parameter. The net effect is "integral symbol becomes smaller" descibed in #3114.

The LaTeX world also has such workaround. I personally don't like this hack and I also don't like this font! Maybe we can have better solution for this.

**Comment by Dherse** [comment]
Perhaps a more "sustainable" workaround might be to support these special parameters in OTF math (if at all possible, I'm not well versed in font stuff), that way all MS fonts would just work™

**Comment by Enter-tainer** [comment]
It is somehow supported, but cambira math has its problem. You may check the long story at #3114.

**Comment by laurmaedje** [comment]
I think WebKit's solution is more elegant. It doesn't need special casing based on the font name.

File: crates/typst/src/math/ctx.rs [hunk 1]
```diff
@@ -202,7 +204,16 @@ impl<'a, 'b, 'v> MathContext<'a, 'b, 'v> {
             let class = self.style.class.as_custom().or(glyph.class);
             if class == Some(MathClass::Large) {
                 let mut variant = if self.style.size == MathSize::Display {
-                    let height = scaled!(self, display_operator_min_height);
+                    let display_operator_min_height = {
+                        // https://github.com/w3c/mathml-core/issues/126#issuecomment-1425804878
+                        let mut height = self.constants.display_operator_min_height();
+                        let min_height = (SQRT_2 * self.ttf.height() as f64) as u16;
```
Comment (Enter-tainer) [inline crates/typst/src/math/ctx.rs:210]: disclaimer: I dont really understant what `self.ttf.height()` is. I also dont understant what is `baseGlyph` in the webkit algorithm.

File: tests/ref/math/class.png
Comment (Enter-tainer) [inline tests/ref/math/class.png:1]: idk why this is changed

Comment (WannesMalfait) [reply tests/ref/math/class.png:1]: I don't know much about the code, but it looks like the only change is to the `a class("large", \/) b` line. The new calculation for the `min_height` gives a bigger height than previously.

File: crates/typst/src/math/ctx.rs (hunk 1 above)
Comment (laurmaedje) [reply crates/typst/src/math/ctx.rs:210]: `self.ttf.height()` isn't the correct metric here. it is a font-global height. what the webkit algorithm is doing is taking the height of the concrete glyph (in the bug case the integral). that's the `baseGlyph` ("base" because it's not yet stretched) and it corresponds to the `glyph` variable in our code. so in this case `self.ttf.height()` needs to be replaced by `glyph.height()`. since the glyph's height is already stretched, I think we can then just take the max of `SQRT_2 * glyph.height()` and the old `scaled!(self, display_operator_min_height)` code.

Comment (Enter-tainer) [reply crates/typst/src/math/ctx.rs:210]: Thanks for explaination! I really have trouble understand different metrics here. :rofl:

File: tests/ref/math/class.png
Comment (Enter-tainer) [reply tests/ref/math/class.png:1]: Interestly this is still changed even when I fix the metrics

**Comment by Enter-tainer** [comment]
image

It works!

**Comment by Leedehai** [comment]
For code hygiene I'd suggest adding a comment describing the what & why of this unusual hack [1], but since it's merged already I guess it's too late now :)

[1] Git history is inconvenient (unless one really sets out to find the commit). For example, the blame view of a large chunk of code was obscured by that "Merge `typst` and `typst-library`" commit.

**Comment by Enter-tainer** [comment]
>  For example, the blame view of a large chunk of code was obscured by that "Merge typst and typst-library" commit.

this can be somehow improved though. like https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/

**Comment by Enter-tainer** [comment]
i agree comment could be improved
